Add pre-commit hook for hlint

[hlint](https://github.com/ndmitchell/hlint) is a Haskell linting and code suggestion tool. Output looks something like this:

```sh
$ hlint --cross cross1.hs cross2.hs print.hs
print.hs:1:8: Error: Use print
Found:
  putStrLn $ show "Hello"
Why not:
  print "Hello"

cross1.hs:6:9: Warning: Reduce duplication
Found:
  a = 1
  b = 2
  c = 3

Why not:
  Combine with cross2.hs:5:9

2 suggestions
```

@sds How would you suggest we parse and display this kind of output? Should we only consider the first line of each suggestion, ignoring the "Found" and "Why not" stanzas?

--- a/config/default.yml
+++ b/config/default.yml
@@ -225,6 +225,13 @@ PreCommit:
     required_executable: 'grep'
     flags: ['-IHn', "\t"]
 
+  Hlint:
+    enabled: false
+    description: 'Analyzing with hlint'
+    required_executable: 'hlint'
+    install_command: 'cabal install hlint'
+    include: '**/*.hs'
+
   HtmlHint:
     enabled: false
     description: 'Analyzing with HTMLHint'

--- a/lib/overcommit/hook/pre_commit/hlint.rb
+++ b/lib/overcommit/hook/pre_commit/hlint.rb
@@ -1,0 +1,32 @@
+module Overcommit::Hook::PreCommit
+  # Runs `hlint` against any modified Haskell files.
+  #
+  # @see https://github.com/ndmitchell/hlint
+  class Hlint < Base
+    MESSAGE_REGEX = /
+      ^(?<file>(?:\w:)?[^:]+)
+      :(?<line>\d+)
+      :\d+
+      :\s*(?<type>\w+)
+    /x
+
+    MESSAGE_TYPE_CATEGORIZER = lambda do |type|
+      type.include?('W') ? :warning : :error
+    end
+
+    def run
+      result = execute(command, args: applicable_files)
+      return :pass if result.success?
+
+      raw_messages = result.stdout.split("\n").grep(MESSAGE_REGEX)
+
+      # example message:
+      #   path/to/file.hs:1:0: Error: message
+      extract_messages(
+        raw_messages,
+        MESSAGE_REGEX,
+        MESSAGE_TYPE_CATEGORIZER
+      )
+    end
+  end
+end

--- a/spec/overcommit/hook/pre_commit/hlint_spec.rb
+++ b/spec/overcommit/hook/pre_commit/hlint_spec.rb
@@ -1,0 +1,58 @@
+require 'spec_helper'
+
+describe Overcommit::Hook::PreCommit::Hlint do
+  let(:config)  { Overcommit::ConfigurationLoader.default_configuration }
+  let(:context) { double('context') }
+  subject { described_class.new(config, context) }
+
+  let(:result) { double('result') }
+
+  before do
+    subject.stub(:applicable_files).and_return(%w[file1.hs file2.hs])
+    subject.stub(:execute).and_return(result)
+  end
+
+  context 'when hlint exits successfully' do
+    before do
+      result.stub(success?: true, stdout: '')
+    end
+
+    it { should pass }
+  end
+
+  context 'when hlint exits unsucessfully' do
+    let(:result) { double('result') }
+
+    before do
+      result.stub(:success?).and_return(false)
+    end
+
+    context 'and it reports a warning' do
+      before do
+        result.stub(:stdout).and_return(normalize_indent(<<-OUT))
+          file1.hs:22:16: Warning: Use const
+          Found:
+            \\ _ -> False
+          Why not:
+            const False
+        OUT
+      end
+
+      it { should warn }
+    end
+
+    context 'and it reports an error' do
+      before do
+        result.stub(:stdout).and_return(normalize_indent(<<-OUT))
+          file1.hs:22:5: Error: Redundant lambda
+          Found:
+            nameHack = \\ _ -> Nothing
+          Why not:
+            nameHack _ = Nothing
+        OUT
+      end
+
+      it { should fail_hook }
+    end
+  end
+end

